### PR TITLE
TCG yugioh search

### PR DESCRIPTION
**Pull request by bumbadadabum**
Made a command searching the yugioh wiki for yugioh cards. This does require a new library for XMLHttpRequest.

### CODE REVIEW

**Comment by codelegend** [comment]
This can be done with a http get.  ie. ``require('http').get(...)``

File: chat-plugins/tcg.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+* This is a command that allows users to search the yugioh wiki for cards. It will display the closest match with a given query, or a separate message if there isn't anything found.
+*/
+
+exports.commands = {
+	ygo: 'yugioh',
+	yugioh: function (target, room, user) {
+		if (room.id !== 'tcg') return this.sendReply("This command can only be used in the TCG room.");
+		if (!this.canBroadcast()) return;
+
+		// The result is fetched from the wikia api with an XMLHttpRequest.
```
Comment (Zarel) [inline chat-plugins/tcg.js]: Why the external dependency?

Doesn't Node have built-in support for GET requests? Just use that. -_-

Comment (Zarel) [inline chat-plugins/tcg.js]: http://nodejs.org/api/http.html#http_http_get_options_callback

Comment (Zarel) [inline chat-plugins/tcg.js]: The amusing thing is, it's even easier to use than XMLHttpRequest. Because XMLHttpRequest sucks.

**Comment by Zarel** [comment]
Also haven't you heard of squashing?

**Comment by Zarel** [comment]
There should not be fix commits in a pull request unless you're fixing things that are broken in the official repository's master.

**Comment by bumbadadabum** [comment]
>Doesn't Node have built-in support for GET requests? Just use that. -_-
Welp there it shows that I have no experience with node, suck at git, and probably js. I'll try to fix it.

**Comment by Joimer** [comment]
Read this when you're done with those issues, will come in handy:
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

Each commit is a logical change, fix commits should be squashed in.

**Comment by bumbadadabum** [comment]
It works now, using the built-in nodejs libraries. Many thanks to ascriptmaster and codelegend.

**Comment by Joimer** [comment]
Travis build does not pass, please do fix:
/home/travis/build/Zarel/Pokemon-Showdown/chat-plugins/tcg.js
line 24 col 21 'result' is not defined.
line 26 col 25 'result' is not defined.
line 30 col 109 'result' is not defined.
line 30 col 137 'result' is not defined.

You need to npm install gulp so you can run npm test on your Showdown before pull requesting.

File: chat-plugins/tcg.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+					if (result.exception) {
+						this.sendReply("No articles matching your query found.");
+					}
+					else {
+						this.sendReplyBox("<strong>Best result for " + query + ":</strong><br/><a href= " + result.items[0].url + ">" + result.items[0].title + "</a>");
+					}
+
+					room.update();
+				}).bind(self));
```
Comment (codelegend) [inline chat-plugins/tcg.js]: why bind? just call ``self.sendReply``

**Comment by bumbadadabum** [comment]
>Travis build does not pass, please do fix:

Yeah I did that now.

File: chat-plugins/tcg.js (hunk 2 above)
Comment (bumbadadabum) [inline chat-plugins/tcg.js]: ah i'll fix that thank you. It was mostly still a remnant of the "why doesn't this work gdi"

**Comment by bumbadadabum** [comment]
And apologies for the need for many "please squash" and "please test"s, I'm too used to making "fuck shit forgot to test" commits for my project.

File: chat-plugins/tcg.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+* This is a command that allows users to search the yugioh wiki for cards. It will display the closest match with a given query, or a separate message if there isn't anything found.
+*/
+
+exports.commands = {
+	ygo: 'yugioh',
+	yugioh: function (target, room, user) {
+		if (room.id !== 'tcg') return this.sendReply("This command can only be used in the TCG room.");
+		if (!this.canBroadcast()) return;
+
+		var http = require("http");
```
Comment (Zarel) [inline chat-plugins/tcg.js]: Move this up to the top of the file.

File: chat-plugins/tcg.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+		var query = Tools.escapeHTML(target);
+
+		var Search = http.get(('http://yugioh.wikia.com/api/v1/Search/List/?query=' + query + '&limit=1'), function (a) {
+			var response = '';
+			a.on('data', function (data) {
+				response += data;
+			});
+			a.on('end', function () {
+				var result = JSON.parse(response);
```
Comment (Slayer95) [inline chat-plugins/tcg.js]: We aren't guaranteed that this won't throw, are we?

File: chat-plugins/tcg.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+				if (result.exception) {
+					self.sendReply("No articles matching your query found.");
+				} else {
+					self.sendReplyBox("<strong>Best result for " + query + ":</strong><br/><a href= " + result.items[0].url + ">" + result.items[0].title + "</a>");
```
Comment (Slayer95) [inline chat-plugins/tcg.js]: This needs escaping.

Comment (Zarel) [inline chat-plugins/tcg.js]: Oh, can you fix those two things, @Slayer95?

Comment (bumbadadabum) [inline chat-plugins/tcg.js]: I don't think it needs escapting to be honest. The query is already escaped, and the results are wikia urls. Unless you want to make a page on the yugioh wiki specifically to mess with PS, that won't be a problem. If someone actually makes such a page it'd be gone within hours anyway.

Comment (Joimer) [inline chat-plugins/tcg.js]: There is a minimal risk -> risk must be fixed. 0.00001 is still >0.

Comment (Zarel) [inline chat-plugins/tcg.js]: > Unless you want to make a page on the yugioh wiki specifically to mess with PS, that won't be a problem.

Yes, unless you want to hack PS, PS won't be hacked. The issue is that there are people who _do_ want to hack PS. 9.9

Comment (Zarel) [inline chat-plugins/tcg.js]: It's like saying "You don't need to lock the door. Unless someone specifically wants to open the door, that won't be a problem."

Actually, the problem is that lots of people want to open the door.

**Pull request merged**